File: server/src/main/java/org/metalscraps/eso/lang/server/ServerMain.java
package org.metalscraps.eso.lang.server;

import com.google.api.client.googleapis.auth.oauth2.GoogleCredential;
import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport;
import com.google.api.client.http.HttpTransport;
import com.google.api.client.json.JsonFactory;
import com.google.api.client.json.jackson2.JacksonFactory;
import com.google.api.services.compute.Compute;
import com.google.api.services.compute.model.Operation;
import org.metalscraps.eso.lang.lib.AddonManager;
import org.metalscraps.eso.lang.lib.bean.ToCSVConfig;
import org.metalscraps.eso.lang.lib.config.AppWorkConfig;
import org.metalscraps.eso.lang.lib.util.Utils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.security.GeneralSecurityException;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.function.Predicate;

class ServerMain {

    private static final Logger logger = LoggerFactory.getLogger(ServerMain.class);
    private final AppWorkConfig appWorkConfig = new AppWorkConfig();
    private final Properties properties = Utils.setConfig(Paths.get("."), Paths.get(".config"), Map.of());

    private void run() {

        logger.info(appWorkConfig.getDateTime().format(DateTimeFormatter.ofPattern("yy-MM-dd hh:mm:ss"))+" / 작업 시작");
        appWorkConfig.setBaseDirectoryToPath(Paths.get(properties.getProperty("WORK_DIR")));
        appWorkConfig.setPODirectoryToPath(appWorkConfig.getBaseDirectoryToPath().resolve("PO_"+appWorkConfig.getToday()));

        // 이전 데이터 삭제
        logger.info("이전 데이터 삭제");
        deletePO();
        logger.info("PO 다운로드");
        Utils.downloadPOs(appWorkConfig);
        logger.info("다운로드 된 PO 파일 문자셋 변경");
        Utils.convertKO_PO_to_CN(appWorkConfig);
        logger.info("CSV 생성");
        makeCSV();

        // 데스티네이션
        new AddonManager(appWorkConfig).destination();

        logger.info("인스턴스 시작");
        var res = startCompressServer();
        logger.info(Objects.requireNonNull(res).getStatus());
    }

    public static void main(String[] args) {
        new ServerMain().run();
    }

    private void deletePO() {

        var workDir = appWorkConfig.getBaseDirectoryToPath()+"/PO_";

        Predicate<Path> p = x -> x.toString().startsWith(workDir) && !x.toString().startsWith(workDir+appWorkConfig.getToday());
        try {
            // 디렉토리 사용중 오류, 파일 먼저 지우고 디렉토리 지우기
            Files.walk(appWorkConfig.getBaseDirectoryToPath()).filter(p.and(Files::isRegularFile)).forEach(path -> { try { Files.delete(path); } catch (IOException e) { e.printStackTrace(); } });
            Files.walk(appWorkConfig.getBaseDirectoryToPath()).filter(p.and(Files::isDirectory)).forEach(path -> { try { Files.delete(path); } catch (IOException e) { e.printStackTrace(); } });
        } catch (IOException e) {
            logger.error(e.getMessage()+" 이전 파일 삭제 실패");
            e.printStackTrace();
        }

    }

    private void makeCSV() {


        var listFiles = Utils.listFiles(appWorkConfig.getPODirectoryToPath(), "po2");
        var fileList = new ArrayList<File>();
        listFiles.forEach(e->fileList.add(e.toFile()));
        var list = Utils.getMergedPO(fileList);
        var config = new ToCSVConfig().setWriteSource(false);

        Utils.makeCSVwithLog(appWorkConfig.getPODirectoryToPath().resolve("kr.csv"), config, list);
        Utils.makeCSVwithLog(appWorkConfig.getPODirectoryToPath().resolve("kr_beta.csv"), config.setBeta(true), list);
        Utils.makeCSVwithLog(appWorkConfig.getPODirectoryToPath().resolve("tr.csv"), config.setWriteFileName(true).setBeta(false), list);
    }

    private Operation startCompressServer() {

        try {
            HttpTransport httpTransport = GoogleNetHttpTransport.newTrustedTransport();
            JsonFactory jsonFactory = JacksonFactory.getDefaultInstance();

            GoogleCredential credential = GoogleCredential.fromStream(new FileInputStream(Path.of(properties.getProperty("GCP_PERM_JSON")).toFile()));
            if (credential.createScopedRequired()) credential = credential.createScoped(Collections.singletonList("https://www.googleapis.com/auth/cloud-platform"));

            Compute computeService = new Compute.Builder(httpTransport, jsonFactory, credential)
                    .setApplicationName("Google-ComputeSample/0.1")
                    .build();

            Compute.Instances.Start request = computeService.instances().start(properties.getProperty("GCP_PROJECT_NAME"), properties.getProperty("GCP_PROJECT_ZONE"), properties.getProperty("GCP_COMPRESS_SERVER_INSTANCE_NAME"));
            return request.execute();
        } catch (IOException | GeneralSecurityException e) {
            e.printStackTrace();
        }
        return null;
    }
}

File: server/src/main/java/org/metalscraps/eso/lang/server/ServerMain.kt
package org.metalscraps.eso.lang.server

import org.metalscraps.eso.lang.lib.AddonManager
import org.metalscraps.eso.lang.lib.config.AppVariables
import org.metalscraps.eso.lang.lib.config.ESOMain
import org.metalscraps.eso.lang.lib.util.Utils
import org.metalscraps.eso.lang.server.config.ServerConfig
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.time.format.DateTimeFormatter
import java.util.*
import java.util.function.Predicate

@Component
class ServerMain(private val config:ServerConfig) : ESOMain {
    private final val logger = LoggerFactory.getLogger(ServerMain::class.java)
    private val vars = AppVariables
    private final val lang:Path
    private final val dest:Path
    private final val ttc:Path

    init {
        vars.baseDir = config.workDir
        lang = vars.workDir.resolve("lang_${AppVariables.todayWithYear}.7z")
        dest = vars.workDir.resolve("destinations_${AppVariables.todayWithYear}.7z")
        ttc = vars.workDir.resolve("tamrielTradeCentre_${AppVariables.todayWithYear}.7z")
    }

    fun init() : Boolean {
        vars.run {
            for(x in dirs) if(Files.notExists(x)) Files.createDirectories(x)
            if(config.isLinux) if(Files.notExists(Paths.get("/root/.ssh/id_rsa"))) {
                logger.error("github id_rsa 존재하지 않음.")
                return false
            }
        }
        return true
    }

    override fun start() {
        vars.run {
            var error = false
            logger.info(dateTime.format(DateTimeFormatter.ofPattern("yy-MM-dd hh:mm:ss")) + " / 작업 시작")
            if(!init()) {
                logger.info("초기화 실패")
                System.exit(-1)
            }

            // 이전 데이터 삭제
            logger.info("이전 데이터 삭제")
            deleteOldData()
            logger.info("PO 다운로드")
            try { Utils.downloadPOs() } catch (e: Exception) { logger.error(e.toString()); e.printStackTrace(); error = true }

            addons()
            logger.info("LANG 생성")
            makeLANG()
            if(compress()) sfx()
            else logger.info("SFX 스킵")

            Utils.processRun(workDir, "echo ${Date().time}/$todayWithYear/${Utils.crc32(Paths.get("$lang.exe"))}", ProcessBuilder.Redirect.to(workDir.resolve("version").toFile()))
            if(!error || config.forceUpload) upload()
        }
    }

    private fun addons() {
        // 데스티네이션
        AddonManager().destination()
        AddonManager().destination("kb", beta = true)
        AddonManager().destination("tr", writeFileName = true)
        AddonManager().tamrielTradeCentre()
        AddonManager().tamrielTradeCentre("kb", beta = true)
        AddonManager().tamrielTradeCentre("tr", writeFileName = true)
    }

    private fun compress() : Boolean {
        System.gc() // 외부 프로세스 사용 시 jvm oom이 안뜨므로 명시적 gc
        val needSfx = Files.notExists(lang) or Files.notExists(dest) or Files.notExists(ttc)
        vars.run {
            logger.info("대상 압축")
            //Utils.kt.processRun(workDir, "7za a -m0=LZMA2:d96m:fb64 -mx=5 $lang $workDir/*.lang") // 최대압축/메모리 -1.5G, 아카이브 17mb
            if(Files.notExists(lang)) Utils.processRun(workDir, "7za a -mmt=1 -m0=LZMA2:d32m:fb64 -mx=5 $lang $workDir/*.lang") // 적당히, 메모리 380m, 아카이브 30m, only 1 threads.
            if(Files.notExists(dest)) Utils.processRun(workDir, "7za a -mx=9 $dest $workAddonDir/Destinations/*")
            if(Files.notExists(ttc)) Utils.processRun(workDir, "7za a -mx=9 $ttc $workAddonDir/TamrielTradeCentre/*")
        }
        if(!needSfx) logger.info("압축 스킵")
        return needSfx
    }

    private fun sfx() {
        vars.run {
            logger.info("SFX 생성")
            val sfx = javaClass.classLoader.getResource("./7zCon.sfx").path
            Utils.processRun(workDir, "cat $sfx $lang", ProcessBuilder.Redirect.to(Paths.get("$lang.exe").toFile()))
            Utils.processRun(workDir, "cat $sfx $dest", ProcessBuilder.Redirect.to(Paths.get("$dest.exe").toFile()))
            Utils.processRun(workDir, "cat $sfx $ttc", ProcessBuilder.Redirect.to(Paths.get("$ttc.exe").toFile()))
        }
    }

    private fun upload() {
        vars.run {
            logger.info("목적파일 업로드")
            // 버전 문서
            Utils.processRun(workDir, "chmod 600 /root/.ssh/id_rsa")
            Utils.processRun(workDir, "git init")
            Utils.processRun(workDir, "git add ${workDir.resolve("version")} $lang.exe $dest.exe $ttc.exe po")
            Utils.processRun(workDir, "git commit -m $todayWithYear")
            Utils.processRun(workDir, "git remote add origin git@github.com:Whya5448/EsoKR-LANG.git")
            Utils.processRun(workDir, "git push -u origin master --force")
            Utils.processRun(workDir, "rm .git -rf")
        }
    }

    private fun deleteOldData() {
        vars.run {
            val workDir = "$baseDir/$WORK_DIR_PREFIX"
            val p = Predicate { x:Path -> x.toString().startsWith(workDir) && !x.toString().startsWith("$workDir$today") }
            try {
                // 디렉토리 사용중 오류, 파일 먼저 지우고 디렉토리 지우기
                Files.walk(baseDir).filter(p.and { x -> Files.isRegularFile(x) }).forEach(Files::delete)
                Files.walk(baseDir).filter(p.and { x -> Files.isDirectory(x) }).forEach(Files::delete)
            } catch (e: IOException) {
                logger.error(e.message + " 이전 파일 삭제 실패")
                e.printStackTrace()
            }
        }
    }

    private fun makeLANG() {
        vars.run {
            var isNotExist = false
            for(x in arrayOf("kr.lang", "kb.lang", "tr.lang")) isNotExist = isNotExist || Files.notExists(workDir.resolve(x))
            if(isNotExist) {
                val list = Utils.getPOList(Utils.listFiles(poDir, "po"))

                if(Files.notExists(workDir.resolve("kr.lang"))) Utils.makeLANGwithLog(workDir.resolve("kr.lang"), list)
                if(Files.notExists(workDir.resolve("kb.lang"))) Utils.makeLANGwithLog(workDir.resolve("kb.lang"), list, beta = true)
                if(Files.notExists(workDir.resolve("tr.lang"))) Utils.makeLANGwithLog(workDir.resolve("tr.lang"), list, writeFileName = true)
            }
        }
    }
}
